### PR TITLE
Remote, mergable configuration

### PR DESCRIPTION
**Pull request by esmet**

### CODE REVIEW

File: ankh/main.go [hunk 1]
```diff
@@ -222,14 +223,14 @@ func execute(ctx *ankh.ExecutionContext) {
 
 func main() {
 	app := cli.App("ankh", "Another Kubernetes Helper")
-	app.Spec = "[-v] [--config] [--kubeconfig] [--datadir] [--context] [--set...] [--filter]"
+	app.Spec = "[-v] [--ankhconfig] [--kubeconfig] [--datadir] [--context] [--set...] [--filter]"
```
Comment (esmet) [inline ankh/main.go]: i made a poor decision in an earlier PR to rename `ankhconfig` to `config` - this reverts that decision

File: ankh/main.go [hunk 2]
```diff
@@ -222,14 +224,14 @@ func execute(ctx *ankh.ExecutionContext) {
...
 
 	var (
 		verbose    = app.BoolOpt("v verbose", false, "Verbose debug mode")
 		ankhconfig = app.String(cli.StringOpt{
-			Name:   "config",
+			Name:   "ankhconfig",
 			Value:  path.Join(os.Getenv("HOME"), ".ankh", "config"),
-			Desc:   "The ankh config to use",
+			Desc:   "The ankh config to use. ANKHCONFIG may be set to include a list of ankh configs to merge. Similar behavior to kubectl's KUBECONFIG.",
```
Comment (mmoresco) [inline ankh/main.go]: What's the use case for allowing multiple ankh config files?

Comment (esmet) [reply ankh/main.go]: i'm thinking that operators will be given references to configs from cluster admins/owners. this makes the most sense when used with an http-based or k8s-based config reference

File: config/config.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+type ConfigMap struct {
+	Data map[string]interface{} `yaml:"data"`
+}
+
+func GetAnkhConfig(ctx *ankh.ExecutionContext, configPath string) (ankh.AnkhConfig, error) {
+	ankhConfig := ankh.AnkhConfig{}
+
+	body := []byte{}
+	var err error
+	if strings.HasPrefix(configPath, "http://") || strings.HasPrefix(configPath, "https://") {
```
Comment (jondlm) [inline config/config.go]: Let's use `net/url`s `Parse` to pick apart the URL without as much string manipulation.

```
package main

import (
	"fmt"
	"net/url"
)

func main() {
	parsed, _ := url.Parse("kubecontext://foo/bar/baz")
	fmt.Println(parsed.Scheme)
	fmt.Println(parsed.Host)
	fmt.Println(parsed.Path)
}
```

```
kubecontext
foo
/bar/baz

Program exited.
```
https://play.golang.org/p/y3lF6zUUuTm

Comment (esmet) [reply config/config.go]: good call

File: context/context.go [hunk 4]
```diff
@@ -57,6 +56,7 @@ type Context struct {
 // AnkhConfig defines the shape of the ~/.ankh/config file used for global
 // configuration options
 type AnkhConfig struct {
+	Include                   []string           `yaml:"include"`
```
Comment (jondlm) [inline context/context.go]: Shouldn't this be a private field? IIUC it's filled in by code and not parsed from the yaml.

File: config/config.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+	}
+
+	body := []byte{}
+	if u.Scheme == "http" || u.Scheme == "https" {
+		resp, err := http.Get(configPath)
+		if err != nil {
+			return ankhConfig, fmt.Errorf("Unable to fetch ankh config from URL '%s': %v", configPath, err)
+		}
+		defer resp.Body.Close()
+		if resp.StatusCode != 200 {
```
Comment (chrisdopuch) [inline config/config.go]: Should this check a range of codes, rather than just strictly 200?

File: context/context.go [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+		errors = append(errors, fmt.Errorf("Missing or empty `supported-environment-classes`"))
+	}
+
+	if ankhConfig.SupportedResourceProfiles == nil || len(ankhConfig.SupportedResourceProfiles) == 0 {
+		errors = append(errors, fmt.Errorf("Missing or empty `supported-resource-profiles`"))
+	}
+
+	selectedContext, contextExists := ankhConfig.Contexts[ankhConfig.ContextName]
+	if contextExists == false {
+		errors = append(errors, fmt.Errorf("Context '%s' not found in `contexts`", ankhConfig.ContextName))
```
Comment (chrisdopuch) [inline context/context.go]: you can avoid the else block by returning early here with the errors

File: agent/ankh-agent/values.yaml [hunk 7]
```diff
@@ -45,13 +45,13 @@ ssh:
...
 ankhconfig: |
-  supported-environments: [ "dev", "production" ]
+  supported-environment-classes: [ "dev", "production" ]
   supported-resource-profiles: ["natural", "constrained" ]
   current-context: local
   contexts:
     local:
       kube-context: local
-      environment: dev
+      environment-class: dev
```
Comment (jondlm) [inline agent/ankh-agent/values.yaml]: You did it! I was secretly hoping you'd change all these references!

File: helm/helm.go [hunk 8]
```diff
@@ -147,27 +147,12 @@ func templateChart(ctx *ankh.ExecutionContext, chart ankh.Chart, ankhFile ankh.A
...
 	_, resourceProfilesError := os.Stat(files.AnkhResourceProfilesPath)
 	if resourceProfilesError == nil {
-		if _, err := util.CreateReducedYAMLFile(files.AnkhResourceProfilesPath, currentContext.ResourceProfile); err != nil {
+		if _, err := util.CreateReducedYAMLFile(files.AnkhResourceProfilesPath, context.ResourceProfile); err != nil {
 			return "", fmt.Errorf("unable to process ankh-resource-profiles.yaml file for chart '%s': %v", chart.Name, err)
 		}
 		helmArgs = append(helmArgs, "-f", files.AnkhResourceProfilesPath)
 	}
 
-	// TODO: add validation for secrets
```
Comment (jondlm) [inline helm/helm.go]: Cool, looks like you removed secrets. Agreed on that.

File: helm/helm.go [hunk 9]
```diff
@@ -138,7 +138,7 @@ func templateChart(ctx *ankh.ExecutionContext, chart ankh.Chart, ankhFile ankh.A
 	// Load `values` from chart
 	_, valuesErr := os.Stat(files.AnkhValuesPath)
 	if valuesErr == nil {
-		if _, err := util.CreateReducedYAMLFile(files.AnkhValuesPath, currentContext.Environment); err != nil {
+		if _, err := util.CreateReducedYAMLFile(files.AnkhValuesPath, context.EnvironmentClass); err != nil {
 			return "", fmt.Errorf("unable to process ankh-values.yaml file for chart '%s': %v", chart.Name, err)
```
Comment (jondlm) [inline helm/helm.go]: Really glad we didn't name that file `ankh-environment-values.yaml` cause we don't need to rename it now! 😌

File: config/config.go [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+			return ankhConfig, fmt.Errorf("Non-200 status code when fetching ankh config from URL '%s': %v", configPath, resp.Status)
+		}
+		body, err = ioutil.ReadAll(resp.Body)
+	} else if u.Scheme == "kubecontext" {
+		parts := strings.Split(strings.Trim(u.Path, "/"), "/")
+		if len(parts) != 3 {
+			return ankhConfig, fmt.Errorf("Malformed URL: kubecontext:// schemes require a path that contains `/kubecontext/namespace/name/key`. Found: \"%s\"", u.Path)
```
Comment (jondlm) [inline config/config.go]: If we want to keep this in, we should probably keep it undocumented.

**Pull request merged**